DOC: optimize: add comparison of optimizers to guide (#20040)
scipy/scipy: doc/source/tutorial/optimize.rst

:mod:`scipy.optimize` (can also be found by ``help(scipy.optimize)``).

Optimization Solvers Reference
------------------------------

> Provide an overview of various optimization solvers,which are tools 
> used to find the minimum or maximum of an objective function subject 
> to constraints. These solvers are divided into two main categories: 
> Local Optimization and Global Optimization.Each category has a set of 
> solvers with unique characteristics and capabilities.

Reviewer: ```suggestion
This section provides an overview of the optimization solvers
available in SciPy to assist users  infinding a suitable solver for a given
problem. These solvers are divided into two main categories: 
local optimization and global optimization. Each category has a set of 
solvers with unique characteristics and capabilities.
```

Local Optimization Solvers
^^^^^^^^^^^^^^^^^^^^^^^^^^

Local optimization solvers are algorithms that search for a local minimum, 
starting from an initial point. These solvers are particularly useful 
when the problem is presumed to have a smooth landscape around the initial guess.

**Local Optimization Solvers Table:**

.. list-table::
   :widths: 15 15 15 20 20 15
   :header-rows: 1

   * - Solver
     - Uses Gradient
     - Uses Hessian
     - Bounds Constraints
     - Nonlinear Constraints
     - Utilizes Sparsity
   * - Nelder-Mead
     - 
     - 
     - ✓
     - 
     - 
   * - Powell
     - 
     - 
     - ✓
     - 
     - 
   * - COBYLA
     - 
     - 
     - ✓
     - 
     - 
   * - CG
     - ✓
     - 
     - 
     - 
     - 
   * - BFGS
     - ✓
     - 
     - 
     - 
     - 
   * - Newton-CG
     - ✓
     - ✓
     - 
     - 
     - ✓
   * - dogleg
     - ✓

>      - must be pos. def.

Reviewer: ```suggestion
     - (✓)
```

     - 
     - 
     - 
   * - trust-ncg
     - ✓
     - ✓
     - 
     - 
     - 
   * - trust-krylov
     - ✓
     - ✓
     - 
     - 
     - 
   * - trust-exact
     - ✓
     - ✓
     - 
     - 
     - 
   * - L-BFGS-B
     - ✓
     - 
     - ✓
     - 
     - 
   * - TNC
     - ✓
     - 
     - ✓
     - 
     - 
   * - trust-constr
     - ✓
     - ✓
     - ✓
     - ✓
     - ✓
   * - SLSQP
     - ✓
     - 
     - ✓
     - ✓
     - 

> 
> 

Reviewer: ```suggestion

(✓) = must be positive definite

```

   :alt: "This X-Y plot is a heatmap with the Z value denoted with the lowest points as black and the highest values as white. The image resembles a chess board rotated 45 degrees but heavily smoothed. A red dot is located at many of the minima on the grid resulting from the SHGO optimizer. SHGO shows the global minima as a red X in the top right. A local minima found with dual annealing is a white circle marker in the top left. A different local minima found with basinhopping is a yellow marker in the top center. The code is plotting the differential evolution result as a cyan circle, but it is not visible on the plot. At a glance it's not clear which of these valleys is the true global minima."
   :include-source: 0

> Global Optimization Solvers

Reviewer 2: ```suggestion
Comparison of Optimizer features
----------------------------------

Global Optimizers
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Global optimization solvers aim to find the global minimum over the entire search space,

Reviewer 2: remove L921 and L922, this info is duplicated from elsewhere.

which is especially important when the objective function has multiple local minima.

> **Global Optimization Solvers Table:**

Reviewer 2: ```suggestion
**Comparison of Global Optimizers:**
```

.. list-table::
   :widths: 20 15 15 20 20 15
   :header-rows: 1

   * - Solver
     - Uses Gradient
     - Uses Hessian
     - Bounds Constraints
     - Nonlinear Constraints
     - Utilizes Sparsity
   * - dual_annealing
     - (✓)
     - (✓)
     - ✓
     - 
     - 
   * - differential_evolution
     - 
     - 
     - ✓
     - ✓
     - 
   * - shgo
     - (✓)
     - (✓)
     - ✓
     - ✓
     - 
   * - direct
     - 
     - 
     - ✓
     - 
     - 
   * - basinhopping
     - (✓)
     - (✓)
     - 
     - 
     - 

(✓) = Depending on the chosen local minimizer

> Local optimization

Reviewer 2: remove L971, L972

------------------

> Local Optimization Solvers

Reviewer 2: ```suggestion
Local Optimizers
```

^^^^^^^^^^^^^^^^^^^^^^^^^^

> Local optimization solvers are algorithms that search for a local minimum, 

Reviewer 2: ```suggestion
Local optimizers are algorithms that search for a local minimum, 
```

Reviewer 2: I think this description can be improved.
They're efficient at finding minima, but can get stuck in local minima. The minima that they find can depend on the point the optimizer starts from.

starting from an initial point. These solvers are particularly useful 
when the problem is presumed to have a smooth landscape around the initial guess.

> **Local Optimization Solvers Table:**

Reviewer 2: ```suggestion
**Local Optimization Solver comparison:**
```

.. list-table::
   :widths: 15 15 15 20 20 15
   :header-rows: 1

   * - Solver
     - Uses Gradient

>      - Uses Hessian

Reviewer 2: This table could be fleshed out more w.r.t gradient/hessian information. e.g. LBFGSB uses a gradient, but that gradient can be estimated by the numerical differentiation, OR the user can supply a gradient function.
In comparison Newton-CG definitely requires a gradient function, but can estimate hessian by numerical differentiation or hessian function.
Yet again trust-constr can estimate both gradient and hessian by functions, or by numdiff for gradient and HessianUpdateStrategy for the Hessian.

i.e. there is lots of complexity in what each function can use.

Reviewer: What about a second table for details about derivatives and hessians?

Author: > What about a second table for details about derivatives and hessians?

@andyfaff

   :alt: "This X-Y plot is a heatmap with the Z value denoted with the lowest points as black and the highest values as white. The image resembles a chess board rotated 45 degrees but heavily smoothed. A red dot is located at many of the minima on the grid resulting from the SHGO optimizer. SHGO shows the global minima as a red X in the top right. A local minima found with dual annealing is a white circle marker in the top left. A different local minima found with basinhopping is a yellow marker in the top center. The code is plotting the differential evolution result as a cyan circle, but it is not visible on the plot. At a glance it's not clear which of these valleys is the true global minima."
   :include-source: 0

> Comparison of Global Optimizers

Reviewer 3: ```suggestion
Global optimization solver comparison
```
To match "Local minimization solver comparison"
But if that's the only thing, I won't change it here.

     - ✓
     - ✓
     - 
     - 
     - 
   * - trust-krylov
     - ✓
     - ✓
     - 
     - 
     - 
   * - trust-exact
     - ✓
     - ✓
     - 
     - 
     - 
   * - L-BFGS-B
     - ✓
     - 
     - ✓
     - 
     - 
   * - TNC
     - ✓
     - 
     - ✓
     - 
     - 
   * - trust-constr
     - ✓
     - ✓
     - ✓
     - ✓
     - ✓
   * - SLSQP
     - ✓
     - 
     - ✓
     - ✓
     - 

Global Optimization Solvers
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Global optimization solvers aim to find the global minimum over the entire search space,
which is especially important when the objective function has multiple local minima.

**Global Optimization Solvers Table:**

.. list-table::
   :widths: 20 15 15 20 20 15
   :header-rows: 1

   * - Solver
     - Uses Gradient
     - Uses Hessian
     - Bounds Constraints
     - Nonlinear Constraints
     - Utilizes Sparsity
   * - dual_annealing
     - (✓)
     - (✓)
     - ✓
     - 
     - 
   * - differential_evolution
     - 
     - 
     - ✓
     - ✓
     - 
   * - shgo
     - (✓)
     - (✓)
     - ✓
     - ✓
     - 
   * - direct
     - 
     - 
     - ✓
     - 
     - 
   * - basinhopping
     - (✓)
     - (✓)
     - 
     - 
     - 

> (✓): Depending on the chosen local minimizer

Reviewer 4: I guess you have a minor doc build issue in CI: `/home/circleci/repo/doc/source/tutorial/optimize.rst:1272: WARNING: Explicit markup ends without a blank line; unexpected unindent.`

There are a few places in the first paragraph above where spaces are missing around punctuation, but that's pretty minor too.

Author: Thank you for your review. Fixed.
